src/System.Private.CoreLib/shared/System/Globalization/DateTimeParse.cs

```diff
@@ -5110,7 +5111,7 @@ private static Exception GetDateTimeParseException(ref DateTimeResult result)
                     return new FormatException(SR.Format(SR.GetResourceString(result.failureMessageID), new string(result.originalDateTimeString), result.failureMessageFormatArgument));
                 default:
                     Debug.Fail("Unknown DateTimeParseFailure: " + result.failure.ToString());
-                    return null;
+                    return null!;
```

Comment:
What does `null!` mean??!

Reply:
I found it weird too since I initially thought it means `tell compiler this is not null` but this actually means `I know what I'm doing, shut up`.

This method generally never returns null except for unexpected case which is prepended with `Debug.Fail`. To tell compiler to not complain I added `!`.

Reply:
What would happen in release mode? It will return null and consuming code would fail with a NRE?

Reply:
it's unexpected by the code so I guess NRE is acceptable? don't have a strong opinion on what we should do in an unexpected case but returning non-nullable seems reasonable if null is not expected

Reply:
I think this is the same as https://github.com/dotnet/coreclr/pull/23539#discussion_r270432973; and we should probably determine what the recommended guidelines are for this type of thing.

Reply:
Look at: https://github.com/dotnet/coreclr/pull/23539#discussion_r270607815